pkg/agent/controller/traceflow/packetin.go

```diff
@@ -409,3 +465,24 @@ func parseCapturedPacket(pktIn *ofctrl.PacketIn) *crdv1alpha1.Packet {
 	}
 	return &capturedPacket
 }
+
+func getEgressObservation(isEgressNode bool, egressIP, egressName string) *crdv1alpha1.Observation {
+	ob := new(crdv1alpha1.Observation)
+	ob.Component = crdv1alpha1.ComponentEgress
+	ob.EgressIP = egressIP
+	ob.Egress = egressName
+	if isEgressNode {
+		ob.Action = crdv1alpha1.ActionMarkedForSNATByHost
+	} else {
+		ob.Action = crdv1alpha1.ActionForwardedToEgressNode
+	}
+	return ob
+}
+
+func getEgressName(c *Controller, ns, pod string) (string, error) {
+	return c.egressQuerier.GetEgress(ns, pod)
+}
+
+func getEgressIPByMark(c *Controller, pktMark uint32) (string, error) {
+	return c.egressQuerier.GetEgressIPByMark(pktMark)
+}
```

Comment:
egressQuerier is already an interface which is quite friendly to mock. You could declare a fakeEgressQuerier in test:
```
type fakeEgressQuerier struct {
    egressName string
    egressIP string
}

func (f *fakeEgressQuerier) GetEgressIPByMark(mark uint32) (string, error) {
    return f.egressName, nil
}

func (f *fakeEgressQuerier) GetEgress(podNamespace, podName string) (string, error) {
    return f.egressIP, nil
}

...
tfc.egressQuerier = &fakeEgressQuerier{egressName: ..., egressIP: ...}
```

Reply:
done, thanks!